File: portal/backend/common/tabular.py
# -*- coding: utf-8 -*-
"""
表格文件读写工具（CSV / XLS / XLSX）
"""
from __future__ import annotations

import csv
import re
from datetime import date, datetime
from io import BytesIO, StringIO
from typing import Iterable, List, Tuple

from flask import Response


# 导入文件大小上限（5MB）
MAX_TABLE_FILE_BYTES = 5 * 1024 * 1024

# CSV/Excel 公式注入防护：单元格以 = + @ 或 制表符/回车 开头，或 - 后跟非数字时加前缀
_FORMULA_RE = re.compile(r'^[=@+\t\r]|^-(?![0-9.])')


def _sanitize_formula(value):
    """防 CSV 公式注入：把可能被 Excel 解释为公式的单元格加单引号前缀"""
    text = str(value)
    if _FORMULA_RE.match(text):
        return "'" + text
    return text


SUPPORTED_TABLE_FILE_TYPES = ('csv', 'xls', 'xlsx')

_MIME_MAP = {
    'csv': 'text/csv; charset=utf-8',
    'xls': 'application/vnd.ms-excel',
    'xlsx': 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
}


def normalize_table_file_type(raw_value, default='csv'):
    value = str(raw_value or '').strip().lower()
    if value in SUPPORTED_TABLE_FILE_TYPES:
        return value
    return default


def infer_table_file_type(filename):
    name = str(filename or '').strip().lower()
    if not name or '.' not in name:
        return None
    ext = name.rsplit('.', 1)[-1]
    if ext in SUPPORTED_TABLE_FILE_TYPES:
        return ext
    return None


def _format_cell_value(value):
    if value is None:
        return ''
    if isinstance(value, datetime):
        return value.strftime('%Y-%m-%d %H:%M:%S')
    if isinstance(value, date):
        return value.strftime('%Y-%m-%d')
    if isinstance(value, float) and value.is_integer():
        return str(int(value))
    return str(value).strip()


def _format_xls_cell(cell_type, cell_value, datemode):
    if cell_value is None:
        return ''
    if cell_type == 3:  # XL_CELL_DATE
        try:
            import xlrd
            value = xlrd.xldate_as_datetime(cell_value, datemode)
            return value.strftime('%Y-%m-%d %H:%M:%S')
        except Exception:
            return ''
    if isinstance(cell_value, float) and cell_value.is_integer():
        return str(int(cell_value))
    return str(cell_value).strip()


def _clean_header_row(headers: Iterable):
    cleaned = []
    for header in headers:
        cleaned.append(str(header or '').strip())
    return cleaned


def _collect_row_map(headers, values):
    row_map = {}
    has_data = False
    for index, header in enumerate(headers):
        if not header:
            continue
        value = values[index] if index < len(values) else ''
        text = _format_cell_value(value)
        row_map[header] = text
        if text != '':
            has_data = True
    return row_map, has_data


def read_table_file(file_storage):
    """
    读取上传文件并返回:
    - fieldnames: 表头数组
    - rows_with_line: [(line_number, row_dict), ...]
    - file_type: csv/xls/xlsx
    """
    if not file_storage:
        raise ValueError('请上传导入文件')

    file_type = infer_table_file_type(file_storage.filename)
    if not file_type:
        raise ValueError('仅支持 csv/xls/xlsx 文件')

    content = file_storage.read(MAX_TABLE_FILE_BYTES + 1)
    if len(content) > MAX_TABLE_FILE_BYTES:
        raise ValueError('文件过大，最大支持 5MB')
    if not content:
        raise ValueError('导入文件内容为空')

    if file_type == 'csv':
        try:
            text = content.decode('utf-8-sig')
        except UnicodeDecodeError as exc:
            raise ValueError('CSV 编码错误，请使用 UTF-8 编码') from exc

        reader = csv.DictReader(StringIO(text))
        fieldnames = [str(item or '') for item in (reader.fieldnames or [])]
        rows = []
        line = 1
        for row in reader:
            line += 1
            normalized_row = {k: ('' if v is None else str(v).strip()) for k, v in (row or {}).items()}
            if not any(str(value).strip() for value in normalized_row.values()):
                continue
            rows.append((line, normalized_row))
        return fieldnames, rows, file_type

    if file_type == 'xlsx':
        try:
            from openpyxl import load_workbook
        except ImportError as exc:
            raise RuntimeError('缺少 openpyxl 依赖，无法处理 xlsx 文件') from exc

        workbook = load_workbook(BytesIO(content), data_only=True, read_only=True)
        sheet = workbook.active
        rows_iter = sheet.iter_rows(values_only=True)

        try:
            header_row = next(rows_iter)
        except StopIteration:
            return [], [], file_type

        headers = _clean_header_row([_format_cell_value(item) for item in header_row])
        rows = []
        for line_number, row_values in enumerate(rows_iter, start=2):
            values = list(row_values or [])
            row_map, has_data = _collect_row_map(headers, values)
            if has_data:
                rows.append((line_number, row_map))
        return headers, rows, file_type

    # xls
    try:
        import xlrd
    except ImportError as exc:
        raise RuntimeError('缺少 xlrd 依赖，无法处理 xls 文件') from exc

    workbook = xlrd.open_workbook(file_contents=content)
    sheet = workbook.sheet_by_index(0)
    if sheet.nrows <= 0:
        return [], [], file_type

    headers = _clean_header_row([
        _format_xls_cell(sheet.cell_type(0, col), sheet.cell_value(0, col), workbook.datemode)
        for col in range(sheet.ncols)
    ])

    rows = []
    for row_index in range(1, sheet.nrows):
        raw_values = [
            _format_xls_cell(sheet.cell_type(row_index, col), sheet.cell_value(row_index, col), workbook.datemode)
            for col in range(sheet.ncols)
        ]
        row_map, has_data = _collect_row_map(headers, raw_values)
        if has_data:
            rows.append((row_index + 1, row_map))
    return headers, rows, file_type


def build_table_response(headers: List[str], rows: Iterable[Iterable], base_filename: str, file_type='csv'):
    file_type = normalize_table_file_type(file_type, default='csv')
    safe_headers = [_sanitize_formula(item) for item in (headers or [])]
    safe_rows = [[_sanitize_formula(_format_cell_value(value)) for value in (row or [])] for row in rows]

    if file_type == 'csv':
        stream = StringIO()
        writer = csv.writer(stream)
        writer.writerow(safe_headers)
        for row in safe_rows:
            writer.writerow(row)
        payload = ('\ufeff' + stream.getvalue()).encode('utf-8')
    elif file_type == 'xlsx':
        try:
            from openpyxl import Workbook
        except ImportError as exc:
            raise RuntimeError('缺少 openpyxl 依赖，无法导出 xlsx 文件') from exc

        workbook = Workbook()
        sheet = workbook.active
        sheet.append(safe_headers)
        for row in safe_rows:
            sheet.append(row)
        buffer = BytesIO()
        workbook.save(buffer)
        payload = buffer.getvalue()
    else:
        try:
            import xlwt
        except ImportError as exc:
            raise RuntimeError('缺少 xlwt 依赖，无法导出 xls 文件') from exc

        workbook = xlwt.Workbook(encoding='utf-8')
        sheet = workbook.add_sheet('Sheet1')
        for col, value in enumerate(safe_headers):
            sheet.write(0, col, value)
        for row_index, row in enumerate(safe_rows, start=1):
            for col_index, value in enumerate(row):
                sheet.write(row_index, col_index, value)
        buffer = BytesIO()
        workbook.save(buffer)
        payload = buffer.getvalue()

    response = Response(payload, mimetype=_MIME_MAP[file_type])
    response.headers['Content-Disposition'] = f'attachment; filename={base_filename}.{file_type}'
    return response

File: apps/api/src/common/tabular.ts
/**
 * Table file read/write (CSV / XLSX)
 *
 * - `.xls` is not supported: uploading .xls returns an explicit 400; export/template with file_type=xls falls back to the default csv
 * - CSV output uses the excel dialect: `\r\n` line endings, minimal quoting, UTF-8 BOM
 * - Formula injection guard: cells starting with = + @ or tab/CR, or - followed by a non-digit, get a `'` prefix
 * - Import file limit is 5MB
 */

import ExcelJS from 'exceljs'
import { parse as parseCsv } from 'csv-parse/sync'
import type { FastifyReply } from 'fastify'

export const MAX_TABLE_FILE_BYTES = 5 * 1024 * 1024
export const SUPPORTED_TABLE_FILE_TYPES = ['csv', 'xlsx'] as const
export type TableFileType = (typeof SUPPORTED_TABLE_FILE_TYPES)[number]

const FORMULA_RE = /^[=@+\t\r]|^-(?![0-9.])/

const MIME_MAP: Record<TableFileType, string> = {
  csv: 'text/csv; charset=utf-8',
  xlsx: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
}

/** Read/validation failure; the caller turns it into a 400 */
export class TableFileError extends Error {}

export interface UploadedFile {
  filename: string
  data: Buffer
}

export type CellValue = string | number | boolean | null | undefined | Date

export function sanitizeFormula(value: unknown): string {
  const text = String(value)
  return FORMULA_RE.test(text) ? `'${text}` : text
}

export function normalizeTableFileType(raw: unknown, fallback: TableFileType = 'csv'): TableFileType {
  const value = String(raw ?? '')
    .trim()
    .toLowerCase()
  return (SUPPORTED_TABLE_FILE_TYPES as readonly string[]).includes(value) ? (value as TableFileType) : fallback
}

function extensionOf(filename: string | null | undefined): string | null {
  const name = String(filename ?? '')
    .trim()
    .toLowerCase()
  if (!name || !name.includes('.')) return null
  return name.slice(name.lastIndexOf('.') + 1)
}

export function inferTableFileType(filename: string | null | undefined): TableFileType | null {
  const ext = extensionOf(filename)
  return ext && (SUPPORTED_TABLE_FILE_TYPES as readonly string[]).includes(ext) ? (ext as TableFileType) : null
}

function pad(n: number): string {
  return String(n).padStart(2, '0')
}

/** Cell value → text: null→'', datetime→'YYYY-MM-DD HH:mm:ss', integral float→integer text, everything else stringified and trimmed */
export function formatCellValue(value: unknown): string {
  if (value === null || value === undefined) return ''
  if (value instanceof Date) {
    return (
      `${value.getUTCFullYear()}-${pad(value.getUTCMonth() + 1)}-${pad(value.getUTCDate())} ` +
      `${pad(value.getUTCHours())}:${pad(value.getUTCMinutes())}:${pad(value.getUTCSeconds())}`
    )
  }
  if (typeof value === 'boolean') return value ? 'True' : 'False'
  if (typeof value === 'number') return Number.isInteger(value) ? String(value) : String(value)
  return String(value).trim()
}

// ---------------------------------------------------------------- Read

export interface TableReadResult {
  fieldnames: string[]
  /** [row number, row dict]; row numbers start at 2 (the line after the header) */
  rows: [number, Record<string, string>][]
  fileType: TableFileType
}

function readCsv(content: Buffer): TableReadResult {
  let text: string
  try {
    text = new TextDecoder('utf-8', { fatal: true }).decode(content)
  } catch {
    throw new TableFileError('CSV 编码错误，请使用 UTF-8 编码')
  }
  if (text.charCodeAt(0) === 0xfeff) text = text.slice(1)

  // csv.DictReader: blank lines are skipped (not counted), rows with mismatched column counts are handled leniently
  const records = parseCsv(text, {
    relax_column_count: true,
    relax_quotes: true,
    skip_empty_lines: true,
  }) as string[][]
  const [header = [], ...body] = records
  const fieldnames = header.map((h) => String(h ?? ''))

  const rows: [number, Record<string, string>][] = []
  let line = 1
  for (const record of body) {
    line += 1
    const row: Record<string, string> = {}
    fieldnames.forEach((name, i) => {
      row[name] = String(record[i] ?? '').trim()
    })
    if (!Object.values(row).some((v) => v.trim())) continue
    rows.push([line, row])
  }
  return { fieldnames, rows, fileType: 'csv' }
}

/** exceljs cell value → raw value equivalent to what openpyxl data_only=True reads */
function xlsxCellRaw(value: ExcelJS.CellValue): unknown {
  if (value === null || value === undefined) return null
  if (value instanceof Date) return value
  if (typeof value !== 'object') return value
  if ('result' in value) return xlsxCellRaw((value as ExcelJS.CellFormulaValue).result as ExcelJS.CellValue)
  if ('richText' in value) return (value as ExcelJS.CellRichTextValue).richText.map((t) => t.text).join('')
  if ('text' in value) return (value as ExcelJS.CellHyperlinkValue).text
  if ('error' in value) return (value as ExcelJS.CellErrorValue).error
  return String(value)
}

async function readXlsx(content: Buffer): Promise<TableReadResult> {
  const workbook = new ExcelJS.Workbook()
  try {
    await workbook.xlsx.load(content as unknown as ArrayBuffer)
  } catch {
    throw new TableFileError('xlsx 文件解析失败，请确认文件格式')
  }
  const activeIndex = workbook.views?.[0]?.activeTab ?? 0
  const sheet = workbook.worksheets[activeIndex] ?? workbook.worksheets[0]
  if (!sheet || sheet.rowCount === 0) return { fieldnames: [], rows: [], fileType: 'xlsx' }

  const rowValues = (rowNumber: number): unknown[] => {
    const row = sheet.getRow(rowNumber)
    const values: unknown[] = []
    for (let col = 1; col <= Math.max(row.cellCount, sheet.columnCount); col += 1) {
      values.push(xlsxCellRaw(row.getCell(col).value))
    }
    return values
  }

  const headers = rowValues(1).map((v) => formatCellValue(v).trim())
  const rows: [number, Record<string, string>][] = []
  for (let r = 2; r <= sheet.rowCount; r += 1) {
    const values = rowValues(r)
    const row: Record<string, string> = {}
    let hasData = false
    headers.forEach((header, i) => {
      if (!header) return
      const text = formatCellValue(values[i])
      row[header] = text
      if (text !== '') hasData = true
    })
    if (hasData) rows.push([r, row])
  }
  return { fieldnames: headers, rows, fileType: 'xlsx' }
}

/** Read an uploaded table file; throws TableFileError (→ 400) on validation failure */
export async function readTableFile(file: UploadedFile | null | undefined): Promise<TableReadResult> {
  if (!file) throw new TableFileError('请上传导入文件')
  if (extensionOf(file.filename) === 'xls') {
    throw new TableFileError('不支持 .xls 格式，请另存为 .xlsx 后重新上传')
  }
  const fileType = inferTableFileType(file.filename)
  if (!fileType) throw new TableFileError('仅支持 csv/xlsx 文件')
  if (file.data.length > MAX_TABLE_FILE_BYTES) throw new TableFileError('文件过大，最大支持 5MB')
  if (file.data.length === 0) throw new TableFileError('导入文件内容为空')
  return fileType === 'csv' ? readCsv(file.data) : readXlsx(file.data)
}

// ---------------------------------------------------------------- Write

/** Encode a single CSV field (excel dialect, minimal quoting) */
function csvField(value: string): string {
  return /[",\r\n]/.test(value) ? `"${value.replace(/"/g, '""')}"` : value
}

function csvRow(fields: string[]): string {
  // A row with "a single empty field" is written as "" so it is not read back as a blank line
  if (fields.length === 1 && fields[0] === '') return '""\r\n'
  return `${fields.map(csvField).join(',')}\r\n`
}

export interface TablePayload {
  payload: Buffer
  contentType: string
  filename: string
}

export async function buildTable(
  headers: unknown[],
  rows: Iterable<unknown[]>,
  baseFilename: string,
  fileTypeRaw: unknown = 'csv',
): Promise<TablePayload> {
  const fileType = normalizeTableFileType(fileTypeRaw)
  const safeHeaders = headers.map(sanitizeFormula)
  const safeRows = [...rows].map((row) => (row ?? []).map((v) => sanitizeFormula(formatCellValue(v))))

  let payload: Buffer
  if (fileType === 'csv') {
    const text = [safeHeaders, ...safeRows].map(csvRow).join('')
    payload = Buffer.from(`﻿${text}`, 'utf8')
  } else {
    const workbook = new ExcelJS.Workbook()
    const sheet = workbook.addWorksheet('Sheet')
    // Empty-string cells get no <c> element: exceljs would write '' as an empty-string cell, so convert to null here
    const blankToNull = (row: string[]) => row.map((v) => (v === '' ? null : v))
    sheet.addRow(blankToNull(safeHeaders))
    for (const row of safeRows) sheet.addRow(blankToNull(row))
    payload = Buffer.from(await workbook.xlsx.writeBuffer())
  }
  return { payload, contentType: MIME_MAP[fileType], filename: `${baseFilename}.${fileType}` }
}

/** Send a table file (sets Content-Type / download filename and writes the content) */
export function sendTable(reply: FastifyReply, table: TablePayload): FastifyReply {
  return reply
    .header('Content-Type', table.contentType)
    .header('Content-Disposition', `attachment; filename=${table.filename}`)
    .send(table.payload)
}
